task_processing/plugins/mesos/logging_executor.py

```diff
@@ -0,0 +1,208 @@
+                    ))
+                    break
+
+                offset = offset + log_length
+                # Stop if there is no more data
+                if log_length < TASK_LOG_CHUNK_LEN:
+                    break
+            # Update offset of this stream
+            offsets[f] = offset
+
+        # Update both offsets for the task
+        with self.task_lock:
+            self.running_tasks = self.running_tasks.set(
+                task_id,
+                log_md.set(log_offsets=pmap(offsets)),
+            )
+
+    # process downstream events
+    def event_loop(self):
+        while True:
+            while not self.src_queue.empty():
+                e = self.src_queue.get()
+                self.dest_queue.put(e)
+                self.src_queue.task_done()
+
+                # Record the base log url
+                if e.kind == 'task' and e.platform_type == 'staging':
+                    url = e.raw.offer.url.scheme + '://' + \
+                        e.raw.offer.url.address.ip + ':' + \
+                        str(e.raw.offer.url.address.port)
+                    self.staging_tasks = self.staging_tasks.set(e.task_id, url)
+
+                if e.kind == 'task' and e.platform_type == 'running':
+                    url = self.staging_tasks[e.task_id]
+                    self.staging_tasks = self.staging_tasks.discard(e.task_id)
+
+                    # Simply pass the needed fields and let the logging thread
+                    # to take care of the slow path discovery.
+                    container_id = e.raw.container_status.container_id.value
+                    executor_id = e.raw.executor_id.value
+                    with self.task_lock:
+                        self.running_tasks = self.running_tasks.set(
+                            e.task_id,
+                            LogMetadata(
+                                log_url=url,
+                                container_id=container_id,
+                                executor_id=executor_id
+                            )
+                        )
+
+                # Fetch the last log and remove the entry
+                if e.kind == 'task' and e.terminal:
+                    with self.task_lock:
+                        self.done_tasks = self.done_tasks.append(e.task_id)
+
+            if self.stopping:
+                return
+
+            time.sleep(1)
+
+    def logging_loop(self):
+        while True:
+            # grab logs
+            for task_id in self.running_tasks.keys():
+                self.stream_task_log(task_id)
+
+            while len(self.done_tasks):
+                task_id = self.done_tasks[0]
+                self.stream_task_log(task_id)
+                with self.task_lock:
+                    self.done_tasks = self.done_tasks.remove(task_id)
+                    self.running_tasks = self.running_tasks.discard(task_id)
+
+            if self.stopping:
+                return
+
+            time.sleep(10)
```

Comment:
maybe reduce to sleep(1)? is files/ endpoint generating too much load on mesos?

Reply:
You are right. I wanted to reduce the number of file API hits as much as possible. Mesos API is single threaded. I don't know exactly how expensive the file API is, but we have seen a significant impact of state.json (the difference is just serialization?).

Reply:
ok, let's run with 10s for now and iterate.